Add JUnit reporter

plu opened the pull request:
In #281 we discussed about additional test reporters. Here comes the JUnit one. In `xctool` I saw that they kind of flatten the nested test suites. This looks better on the CI report especially if you have mixed Objective-C and Swift test cases in your test target. But in the end I think supporting nested test suites is the way to go. Maybe I will add a `BOOL flattenTestSuites` later.

plu commented:
I have a bit mixed feelings about the naming of the classes, that I created in this pull request. I believe that following classes might be shared by other test reporters:

* FBTestManagerTestReporterTestCase
* FBTestManagerTestReporterTestCaseFailure
* FBTestManagerTestReporterTestSuite

Hence they are missing the `JUnit` part in their name.

lawrencelomax commented:
😃  Looks like a lot of thought and effort went into this.
Might take a while for me to review this one, but will try to get around to it today.

plu commented:
Thanks! I'm in the middle of the `RSpec` formatter. Maybe it'd make sense for you waiting for that, I noticed it makes sense to extract some common things there.

lawrencelomax commented on XCTestBootstrapTests/FBTestManagerTestReporterJUnitTests.m:
```diff
@@ -0,0 +1,111 @@
+  [[NSFileManager defaultManager] createFileAtPath:self.outputFilePath contents:nil attributes:nil];
+
+  self.outputFileHandle = [NSFileHandle fileHandleForWritingAtPath:self.outputFilePath];
+  self.reporter = [FBTestManagerTestReporterJUnit withOutputFileHandle:self.outputFileHandle];
```
It looks like you never close `outputFileHandle` which means you'll leak file descriptors in these tests.

lawrencelomax replied:
Ignore that, I just looked at the docs and `NSFileHandle` will close the resource when it is deallocated.

lawrencelomax commented on XCTestBootstrap/TestManager/FBTestManagerTestReporterJUnit.h:
```diff
@@ -0,0 +1,30 @@
+
+ @param outputFileHandle a file handle the JUnit XML should be written to.
+ */
++ (instancetype)withOutputFileHandle:(NSFileHandle *)outputFileHandle;
```
Is the main reason for using a file handle here that you can stream the results to a pipe instead of a file?

It does add a burden to the consumer of the API to ensure that the file handle is closed. If you only care about files you could just give a file path and manage the file handle internally to the class. The reporter could also clean the file handle on `dealloc` or when there is a final event in the `FBTestManagerTestReporter` protocol indicating that the stream should be closed.

lawrencelomax replied:
Add an `@return` in the documentation comment.

plu replied:
For the `RSpec` reporter you probably want to write to `stdout` or `stderr`, I assume. I gave the `JUnit` reporter the same interface so later it will be easier to plug that into `fbsimctl`. I had an interface similar to this in mind:

```
fbsimctl --reporter=rspec # writes to stdout
fbsimctl --reporter=junit junit.xml # write to the file junit.xml
```

If all reporters share the same interface (init with a file handle), it will be easier. But on the other hand you will probably never want to write the `JUnit` output to `stdout` or `stderr`, but always to a file. I guess I should change the interface then.

lawrencelomax commented on XCTestBootstrap/TestManager/FBTestManagerTestReporterJUnit.m:
```diff
@@ -0,0 +1,185 @@
+- (void)testManagerMediatorDidFinishExecutingTestPlan:(FBTestManagerAPIMediator *)mediator
+{
+  NSXMLDocument *document = [FBTestManagerTestReporterJUnit documentForTestSuites:self.rootTestSuites];
+  [self.outputFileHandle writeData:[document XMLDataWithOptions:NSXMLNodePrettyPrint]];
```
It looks like you're only using the file handle to write the data in one big batch at the end. If you only want to support writing to file, you can simplify the API by just passing a file path in the constructor and avoid the need to pass a file handle.

lawrencelomax commented on XCTestBootstrap/TestManager/FBTestManagerTestReporterTestCase.h:
```diff
@@ -0,0 +1,64 @@
+/**
+ The test case duration.
+ */
+@property (nonatomic) NSTimeInterval duration;
```
Make this externally `readonly` if it's not for consumers to modify. You can add a method like `finishWithDuration:status:` that can instead be called from `FBTestManagerTestReporterJUnit`. This can place additional requirements that the  `finishWithDuration:status:` is called only once per-instance.

lawrencelomax commented on XCTestBootstrap/TestManager/FBTestManagerTestReporterTestCase.m:
```diff
@@ -0,0 +1,72 @@
+                                    self.statusDescription, self.duration];
+}
+
+- (NSString *)statusDescription
```
Add the inverse of the `+[FBTestManagerResultSummary statusForStatusString:]` method and call it from here.

lawrencelomax commented on XCTestBootstrap/TestManager/FBTestManagerTestReporterTestCase.h:
```diff
@@ -0,0 +1,64 @@
+/**
+ The test report status.
+ */
+@property (nonatomic) FBTestReportStatus status;
```
add `assign`. Also see the `readonly` comment below.

lawrencelomax commented on XCTestBootstrapTests/FBTestManagerTestReporterJUnitTests.m:
```diff
@@ -0,0 +1,111 @@
+#import "FBTestManagerResultSummary.h"
+#import "FBTestManagerTestReporterJUnit.h"
+
+#define testSuiteDidStart(name, startTime)                                                                             \
```
Is there a good reason for using Macros here? You could define each of these as instance methods of the test case and get the same effect. Macros will make this harder to debug when a test fails since all of the parameters for the assertions will be inlined.

lawrencelomax commented:
Having a base `Reporter` model for accumilating test state makes a lot of sense here. There are a few comments that need addressing but the approach looks sound.

You may also want to consider writing an integration or end-to-end test for this. [`FBSimulatorTestInjectionTests`](https://github.com/facebook/FBSimulatorControl/blob/master/FBSimulatorControlTests/Tests/Integration/FBSimulatorTestInjectionTests.m) shows what this make look like. `fbsimctl` also has a suite of full [end-to-end tests in python](https://github.com/facebook/FBSimulatorControl/blob/master/fbsimctl/cli-tests/tests.py) that will allow you a place to assert based on the output. We have a `SimpleTestTarget.xctest` target that's a fixture of the `FBSimulatorControl` project that will provide you a full test target for running against.

plu commented:
Thank you for the detailed review! I will follow up and address all the issues you mentioned.

plu commented:
I have addressed all your comments, thanks again for taking the time for the review :).

When I was running the test suites of different projects I have noticed that there will always be only one root test suite. So I have adopted the implementation to that. The bits of collecting the test result data I have extracted in a common base class called `FBTestManagerTestReporterBase`, which the `RSpec` reporter can inherit from then too.

lawrencelomax commented on XCTestBootstrapTests/FBTestManagerTestReporterJUnitTests.m:
```diff
@@ -0,0 +1,169 @@
+
+- (void)testJUnitReporterWithManyNestedTestSuites
+{
+  [self testSuite:@"One" didStartAt:@"2016-08-07 10:31:33"];
```
👍👍

lawrencelomax commented on XCTestBootstrap/TestManager/FBTestManagerTestReporterTestCase.m:
```diff
@@ -0,0 +1,72 @@
+
+- (void)finishWithStatus:(FBTestReportStatus)status duration:(NSTimeInterval)duration
+{
+  NSAssert(!self.finished, @"finishWithStatus:duration: may be called only once");
```
👍

lawrencelomax commented on XCTestBootstrap/TestManager/FBTestManagerResultSummary.h:
```diff
@@ -31,8 +31,22 @@ NS_ASSUME_NONNULL_BEGIN
+/**
+ Returns a status enum value for the given status string.
+
+ @param statusString the status string.
```
Nice catch!

lawrencelomax commented:
This is really nice work. I know you've seen the `fbxctest` diff that's landed. Obviously we will eventually want to migrate all of this reporting to `fbxctest`, but the work you've done here can be re-used in `fbxctest`.

@facebook-github-bot import

facebook-github-bot-6 commented:
Thanks for importing.If you are an FB employee go to [Phabricator](https://phabricator.intern.facebook.com/D3689915) to review internal test results.

plu commented:
Makes totally sense to have a different CLI for controlling the Simulators and for executing the tests.

Regarding reporters: Next I'll look into the `RSpec` formatter.

lawrencelomax commented:
I've ran the our linter over your PR and there were a few changes. You'll have to rebase when it lands, but there's nothing that should be too surprising.

plu commented:
Don't worry about it! Thanks for reviewing/importing it :)!
